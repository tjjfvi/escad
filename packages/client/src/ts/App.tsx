
import React from "react";
import Disconnected from "./Disconnected";
import Preview from "./Preview";

const App = () => <>
  <Preview/>
  <Disconnected/>
</>

export default App;
