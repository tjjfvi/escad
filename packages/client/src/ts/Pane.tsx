// @ts-nocheck

import React from "react";
import { useObservable } from "rhobo";

const Pane = ({ name, className = name, children, right = false, left = !right }) => {
  let open = useObservable.use(false);
  return <div className={(open() ? "open " : "") + (left ? "left" : "right") + " Pane " + className}>
    <div className="side" onClick={() => open.toggle()}><span>{name}</span></div>
    <div>{children}</div>
  </div>
}

export default Pane;
