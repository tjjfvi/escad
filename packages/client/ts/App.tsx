
import React from "react";
import Disconnected from "./Disconnected";
import Preview from "./Preview";
import Parameters from "./Parameters";
import Hierarchy from "./Hierarchy";

const App = () => <>
  <Parameters />
  <Preview />
  <Hierarchy />
  <Disconnected />
</>

export default App;
