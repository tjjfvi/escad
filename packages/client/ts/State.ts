/* @flow */

import { observable } from "rhobo";
import WS from "./ws";

class State {

  ws = new WS("ws" + window.location.toString().slice(4) + "ws/", this);

  connected = observable<boolean>(false);
  id = observable<string>();
  serverId = observable<string>();
  paramDef = observable<any>([]);
  params = observable<any>({});
  hierarchy = observable<any>();
  shas = observable<Array<string>>([]);

  constructor() {
    // this.ws.on("hierarchy", d => this.hierarchy(this.processHierarchy(d)));
    this.ws.on("shas", d => this.shas(d));
    this.ws.on("paramDef", d => this.paramDef(d));
    this.params.on("update", () => this.ws.s("params", this.params()))
  }

  // processHierarchy({ name, important, shas, children }) {
  //   children = children.map(x => this.processHierarchy(x));
  //   shas = shas.length ? shas : children.flatMap(x => (x.pre[0] || x).shas);
  //   let tree = { children, shas, name, important, pre: [], post: [], meta: {} }
  //   tree.leaf = children.every(c => c.leaf && !c.important);
  //   let child = children.length === 1 ? children[0] : null;
  //   Object.assign(tree,
  //     !tree.leaf && child ?
  //       !child.important ?
  //         { children: child.children, post: [{ name: child.name, shas: child.shas }, ...child.post] } :
  //         important ?
  //           {} :
  //           { ...child, pre: [{ name, shas }, ...child.pre] } :
  //       {}
  //   );
  //   return tree;
  // }

}

export default new State();
export { State };
