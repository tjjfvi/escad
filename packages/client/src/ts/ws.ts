
import { EventEmitter } from "tsee";
import { State } from "./State";
import flatted from "flatted";

class WS extends EventEmitter<any> {

  state: State;
  url: string;
  ws: WebSocket;

  timeout: any;

  constructor(url: string, state: State){
    super();
    this.url = url;
    this.state = state;
    this.ws = this.initWs();
  }

  s(...data: any){
    if(this.ws.readyState !== 1)
      return;
    console.log("→", ...data);
    this.ws.send(flatted.stringify(data));
  }

  initWs(){
    this.timeout = null;

    if(this.ws)
      this.ws.close();

    this.ws = new WebSocket(this.url);
    this.ws.addEventListener("open", () => {
      this.s("init", this.state.id(), this.state.serverId());
    });

    this.ws.addEventListener("error", () => {
      if(!this.timeout) this.timeout = setTimeout(() => this.initWs(), 5000);
    })

    this.ws.addEventListener("message", msg => {
      let [type, ...data] = flatted.parse(msg.data);

      this.state.connected(true);
      clearTimeout(this.timeout);

      this.timeout = setTimeout(() => {
        this.state.connected(false);
        this.ws.close();
        this.initWs();
      }, 5000);

      if(type === "ping")
        return;

      if(type === "id")
        this.state.id(data[0]);

      if(type === "serverId")
        this.state.serverId(data[0]);

      this.emit("message", type, ...data);
      this.emit(type, ...data);

      console.log("←", type, ...data);
    })

    return this.ws;
  }

}

export default WS;
