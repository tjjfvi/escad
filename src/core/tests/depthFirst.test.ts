import { depthFirst, Stack } from "../mod.ts";

test("Stack", () => {
  const stack = new Stack<number>([3, 2]);
  stack.push(1).push(0);
  expect(stack.pop()).toEqual(0);
  expect([...stack]).toEqual([1, 2, 3]);
  expect(stack.pop()).toEqual(undefined);
});

test("depthFirst", () => {
  type Node = number | Node[];
  const nodes: Node[] = [];
  const tree = [[[[5]], 4], [[3], [2, [1], 0]]];
  depthFirst<Node>([tree], function* (x) {
    nodes.push(x);
    yield* x instanceof Array ? x : [];
  });
  expect(nodes).toMatchSnapshot();
});
