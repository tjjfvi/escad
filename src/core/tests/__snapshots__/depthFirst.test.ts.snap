export const snapshot = {};

snapshot[`depthFirst 1`] = `
[
  [
    [
      [
        [
          5,
        ],
      ],
      4,
    ],
    [
      [
        3,
      ],
      [
        2,
        [
          1,
        ],
        0,
      ],
    ],
  ],
  [
    [
      3,
    ],
    [
      2,
      [
        1,
      ],
      0,
    ],
  ],
  [
    2,
    [
      1,
    ],
    0,
  ],
  0,
  [
    1,
  ],
  1,
  2,
  [
    3,
  ],
  3,
  [
    [
      [
        5,
      ],
    ],
    4,
  ],
  4,
  [
    [
      5,
    ],
  ],
  [
    5,
  ],
  5,
]
`;
